### PR TITLE
Backup/Import improvements (series/episode selection + modules fix)

### LINKED ISSUE #193: Improve user experience around the video import wizard
As implemented in #161, the video import wizard lets teachers easily import videos from other courses and especially modify existing LTI modules in the course to point to the imported episode ID afterwards.

However, we have seen things like this happen several times:
* A teacher has a course from the last lecture term where he uploaded many Opencast videos and created Opencast LTI episode modules for these videos.
* A teacher has to fill the course for the next lecture term and simply imports the course content from the old course with the Moodle core course import wizard.
* As the Opencast LTI episode modules are offered within the Moodle core course import wizard, he checks them to be imported and is thinks this is all he needs to do to make the videos avaiable to students again.
* However, as he didn't use the Opencast video import wizard (or does not even know about it) and the videos are not physically in the course's Opencast series, students won't be able to access the videos in the course through LTI. Even worse, due to the way how Opencast LTI works, the teacher _can_ access the videos in the Opencast LTI episodes (as he is a member of the source Moodle course as well) and does not even have a clue why students can't watch the videos in this course.

As we want to avoid teacher frustration and trouble tickets, we think that this user experience should be improved.

Possible solution approaches would be:

1. Semiautomated solutions:
* Add a warning to the Moodle core course import wizard which is shown as soon as Opencast videos (with or without) LTI episodes exist in the course and which tells the teacher that he has to use the Opencast import wizard as well.
* Add a checkbox "Support me to import Opencast videos as well" to the Moodle core course import wizard which redirects to the Opencast import wizard as soon as the Moodle core course import wizard has done his work.
* Show a warning banner on the course main page as soon as orphaned Opencast LTI modules exist in the course and which points the teacher to the Opencast import wizard.
* Show a warning banner on the Opencast overview page as soon as orphaned Opencast LTI modules exist in the course and which points the teacher to the Opencast import wizard.

2. Automated solutions:
* Implement an addition to the Moodle core course import wizard which automagically imports all videos for which Opencast LTI modules are imported and relinks the episode IDs within these modules without asking the teacher anything.

3. Preventing solutions:
* Modify the Moodle core course import wizard in a way that the teacher does not have a possibility to import Opencast LTI modules at all. This way, nothing can go wrong, but the teacher must recreate all modules from scratch in the new course.

4. Informational solutions:
* Modify the Opencast LTI modules in a way that they show a notifications within the description if the video can't be viewed by students of this course.

I am aware that not all approaches can be implemented within block_opencast. Some will require a companion local plugin, some might require core hacks. This is just a brainstorming list.

### PR DESCRIPTION
**Pull request by ferishili**
This PR fixes #193

### Description
Currently there is no possibility to select series and episode in course import wizard when selection "Duplicate Event" as import mode, and then the entire series and its events in a course will be imported. Apart from that, after the import is completed if you have also imported LTI modules or Activity modules, they also get imported in the new course, but they are still pointing to the event or series from the source course that everything has been imported from.

### How it works.
- In course import wizard, (backup stage) you can now select/deselect series/events when "Duplicate event" as import mode is used.
- In course import wizard, (restore stage) there is no selection possibility, why? because having to select from source in the last (backup stage) does the job, and repeating that is an overkill!
- A new db table is introduced to hold the import mapping data called "block_opencast_importmapping"
- (During restore process): we collect mapping records including series and event ids.
- (After restore process):
   - Series will be fixed in this stage, because there is no need to wait for any workflow to do the duplication process.
   - Events: for events now the import mapping gets the flag that indicated the restore process is completed, why? because, we make sure that all modules including LTI and activities are imported.
   - In duplicate event adhoc task, when the workflow is finished successfully, we then record the workflow id in import mapping record of the event and we schedule another additional adhoc task called "process_duplicated_event_module_fix", which is responsible to get the duplicated event id and run the lti and activity module fix.

### How does the fix methods work:
the whole concept is that the newly created course after import contains modules that has series/event id from the old course. Therefore, the only goal is to collect the newly created series id, or duplicated event id and replace them in the modules (if any) of newly created course.

### NOTE
- The PR also includes in-depth fixing module feature in PHPUnit tests (backup_test.php)
- Because we need to fix both LTI and Activity (mod) modules, we need to the fixing here in block_opencast!
- The version is bumped to add new db table as well as new adhoc task.
- In some places, I have added my name as the author, because the whole thing has been change completed in those files/classes.

Please refer to https://github.com/Opencast-Moodle/moodle-block_opencast/issues/193#issuecomment-1958978574
This PR will fix: https://github.com/Opencast-Moodle/moodle-mod_opencast/issues/34 as well

### CODE REVIEW

**Comment by ferishili** [comment]
Hi @mwuttke,
Please test this PR again, and write your feedback, so we can proceed with merging it.
Thanks in advance.

**Review by justusdieckmann — CHANGES_REQUESTED** [review]
Hey @ferishili,

thank you very much!! I'm sorry for the wait, I have been rendered a bit ineffective for the last few weeks by a kidney stone...

I've left some notes, some more important than others :)

Justus

**Comment by ferishili** [comment]
Hi @justusdieckmann,
Sorry to hear that, and hope you've healed and are well now.

Thanks for your review, I will try my best to apply the changes as soon as possible.
Best Regards
Farbod

File: backup/moodle2/backup_opencast_block_task.class.php
Comment (ferishili) [reply backup/moodle2/backup_opencast_block_task.class.php:107]: series and episodes do not need the instance id, as it is included in `$includesettingname`, which works as a wrapper section in a top-level. And at the end there would a single xml file for each instance separately!

Have a look at:
- line ~75 => `$includesettingname`
- Ultimately, line ~213 => the single xml file name for each ocinstance

File: backup/moodle2/restore_opencast_block_task.class.php
Comment (ferishili) [reply backup/moodle2/restore_opencast_block_task.class.php:128]: This is I believe, in order to escape the steps when other modes of course reuse is performed eg. restore etc and that might happen when a restore is using an old backup file and there is no such step defined, therefore, the restore process does not throw error.
And this at the same time makes sure that if you mode import the step has to be there, because import is a sequence of backup then import!

File: classes/task/process_duplicated_event_module_fix.php
Comment (ferishili) [reply classes/task/process_duplicated_event_module_fix.php:43]: Could you please add your suggestion!
And is it something that older versiuon Moodle that the plugin supports could handle as well?

File: tests/backup_test.php
Comment (ferishili) [reply tests/backup_test.php:85]: Thanks for the suggestion, but does that make a difference?
I would prefer to keep that align with apiurl

File: classes/local/importvideosmanager.php
Comment (ferishili) [reply classes/local/importvideosmanager.php:650]: I though that has been taken care of in the actual method activitymodulemanager::fix_imported_series_modules_in_new_course line 653 , where we check if the return value of DB->get_records() is not empty!

File: classes/local/importvideosmanager.php
Comment (ferishili) [reply classes/local/importvideosmanager.php:683]: https://github.com/Opencast-Moodle/moodle-block_opencast/pull/362#discussion_r1591093215

File: backup/moodle2/backup_opencast_block_task.class.php
Comment (justusdieckmann) [reply backup/moodle2/backup_opencast_block_task.class.php:108]: The problem is that while building the backup/restore plan, the task identifies its settings (for example in `get_setting()`) only by name, not by the hierarchy, and it even [throws an error there](https://github.com/moodle/moodle/blob/main/backup/util/plan/base_task.class.php#L92) if it finds two settings with the same name.

I think it is unlikely that two different instances will have the same series identifier, but in case someone duplicates their opencast instance or something like that, we probably want to tread safer by including the instanceid in the settingname, just to guarantee the uniqueness of the settingname

File: classes/local/importvideosmanager.php
Comment (justusdieckmann) [reply classes/local/importvideosmanager.php:650]: The `$DB->get_records`-call uses `opencasttype`, which is defined in `mod_opencast`, but even if this wasn't an issue, it would throw an sql exception because the `opencast` table does not exist without `mod_opencast`, so sadly it needs to be checked manually, for example with `core_plugin_manager::instance()->get_plugin_info('mod_opencast') != null`

File: tests/backup_test.php
Comment (justusdieckmann) [reply tests/backup_test.php:85]: Not really, just that I thought it might generally be good practice to use `const` when the content of variables doesn't/shouldn't change (apiurl could of course be a const to), so that the variable cannot be used incorrectly, and also the IDE would complain instantly. But it also would be okay for me if it stays like this.

File: classes/task/process_duplicated_event_module_fix.php
Comment (justusdieckmann) [reply classes/task/process_duplicated_event_module_fix.php:42]: I added it, but have not yet tested it, because I don't have my test instance setup here. But it should work, and I will definitely test it once more before merging :)

**Comment by justusdieckmann** [comment]
Hi @ferishili,

thank you, I replied to all comments!

Kind Regards
Justus

**Comment by ferishili** [comment]
Hey @justusdieckmann,
the changes are there, thanks a lot.
Hope they work fine, and I did not miss something!

BR

**Comment by ferishili** [comment]
Hey @justusdieckmann, Is there something else left to do here?

**Comment by justusdieckmann** [comment]
Could you look at this in Moodle 4.4? I always get the error 'Your settings have been altered due to unmet dependencies' but can't find the specific cause.

**Comment by ferishili** [comment]
Hi @justusdieckmann,
as I mentioned this in last community meeting, this PR was promised and delivered only for Moodle up to 4.3! 
Let's merge this PR first and then we tackle the issues regarding refactoring for Moodle 4.4 in separate issue.

**Comment by NinaHerrmann** [comment]
Hey @ferishili as said in the meeting we wanted to have the PR for the 4.4 Release. Therefore, we need it to be 4.4 ready.

**Comment by justusdieckmann** [comment]
Hey @ferishili,

if it would just be refactoring and code style fixes, I would agree. I actually was holding off creating and merging the code style fixes for 4.4 in order to not create more hassle and merge conflicts for you, since 4.4 for example wants to have the language strings sorted alphabetically by key.

But this seems to be a functional bug which completely prevents imports in 4.4. Moodle 4.4 has been released for a few weeks now and we want to release a 4.4 version as soon as possible, so if we already found this bug here, it really doesn't make sense to delay fixing it

**Comment by justusdieckmann** [comment]
Hey @ferishili,

I've added a fix, does it look good to you?

**Comment by ferishili** [comment]
Hey @justusdieckmann,
I have tested the fix for Moodle 4.4, unfortunately it throws error stating `error/settings_lock_by_config` in Moodle 4.1!
@mwuttke would you please test this on your test instance (Moodle 4.1 if I am not mistaken)?

**Comment by justusdieckmann** [comment]
Hey @ferishili,

that's weird, I have tested it on Moodle 4.1 and no error occured

**Comment by mwuttke** [comment]
Hello @ferishili,

I was also unable to detect any errors during my tests with Moodle 4.1 of this pull request.

**Comment by ferishili** [comment]
My scenario that end up with that error:
- Latest PR changes
- In a course overview page -> Course reuse
- Import as mode selected
- Select a course that has Opencast Videos with a few published videos
- Click on next
- Error: error/setting_locked_by_config

**Comment by ferishili** [comment]
As video:


https://github.com/Opencast-Moodle/moodle-block_opencast/assets/53179227/02b8d9ba-a16f-45a1-88e7-83751dd1bc97

**Comment by ferishili** [comment]
Not at the moment, let me see what I can do tomorrow!

**Comment by justusdieckmann** [comment]
I've just had a look through the rest of the code, and I think it isn't really a problem with 4.1, but rather the ACL mode. I've pushed a commit and I believe the error should not occur anymore

**Comment by ferishili** [comment]
Hey @justusdieckmann,
thanks for the fix,
However, your fix now ends up providing a logical bug!

By "ACL Change" as for `importmode` and "Unchecked" as for `importvideoscoredefaultvalue`,  the episodes are now getting the RED "X" no matter if you select the series to import or leave it unchecked, which is logically **incorrect** because it creates a big misunderstanding from UI/UX point of view.

IMHO they should always remain checked with green tick to avoid any confusion! The Help tooltip icon is also there to explain the logic!

Your opinions here are most welcome @mwuttke @justusdieckmann

BR

**Comment by mwuttke** [comment]
Hello @ferishili, Hello @justusdieckmann,

I've tested this pull request again by using the "ACL Change" and the "Unchecked" mode. And I think Farbod is right with his comment. We normally use "Duplicating Events". So I had forgotten to look at the other variant.

**Comment by justusdieckmann** [comment]
Hey,

ok, so I'm not completely sure if moodle even supports that kind of dependency, because it seems a dependency always causes the child to be disabled if the parent has a specific value (I couldn't find a dependency without any functional implication), and disabled settings that are turned on initially (or locked in the on state) throw an error.

Honestly, I just didn't want to fight the dependency setting anymore, so I've just listed the videos in the series setting in ACL mode. Does this work for you?
image

**Comment by mwuttke** [comment]
Hello @justusdieckmann.

It's not what I would expect in terms of look and feel, but it does what it's supposed to ;-)

Thanks a lot for your review!

And thanks a lot for you @ferishili for your work.

**Comment by ferishili** [comment]
Hi,
Thanks @justusdieckmann and @mwuttke,
From my side, that looks ok (it is not optimal, but it is acceptable), we could finally get this PR merged!

BR

**Review by justusdieckmann — APPROVED** [review]
Thank you very much!

**Pull request merged**
